TRIVIAGEN-8: Navigation

Kazoroo opened the pull request:
Added navigation in whole app. For the TriviaGameScreen I have created local states to handle the game behavior (I mean that question change), because creating view models was beyond my task. This also refers to quit button. The test doesn't allow to use LocalContext in the composable, so it must be done by creating a special function to quit app in the view model.
 
I don't know if it's shown well in the video, but once the user finishes the trivia, they can't go back.

https://github.com/sidcgithub/ai-trivia-app-android/assets/91783342/feb1e08d-2963-4397-94e4-c04d8bb99b90




One question for reviewer: can the navController parameter in TriviaGenScaffold.kt be optional?

Kazoroo committed:
Setup navigation, add basic navigation for game screen

Navigation is made to the ResultsScreen, changed the size of question in TriviaGameScreen, to do this added new TextStyle to Type.kt, in TriviaGameScreen add glow when user click an option, navigation to next question, and if there is no other questions go to ResultsScreen. Add dependencies for navigation, and update one dependency.

Kazoroo committed:
Add navigation for ResultsScreen and AnswersScreen. Optimize imports in build.gradle.kts

Kazoroo committed:
Move screen routes to a new class Route.kt

Kazoroo committed:
Write few tests for navigation.

Add semantics to screens.

Kazoroo committed:
Code cleanup and refactoring

Kazoroo committed:
Fix typo

sidcgithub commented:
In `TriviaGenScaffold` can we use a Sealed class instead of making navcontroller nullable?

```
sealed class NavigationStatus {
    object None : NavigationStatus()
    data class Enabled(val navController: NavController) : NavigationStatus()
    data class BackDisabled(val navController: NavController) : NavigationStatus()
}
```
For `NavigationStatus` `None` or `BackDisabled` we can disable the back option in the `AppBar`

sidcgithub commented on app/src/main/java/com/triviagenai/triviagen/trivia/presentation/mainmenu/MainMenuScreen.kt:
```diff
@@ -11,20 +11,23 @@ import androidx.compose.material3.TextButton
             modifier = Modifier
-                .fillMaxSize(),
+                .fillMaxSize()
+                .semantics { contentDescription = "MainMenuScreen" },
```
Would it be better to identify using testTag (MainMenuScreen) instead of contentDescription if we're only using the description for testing purposes. If we're using it for both purposes can we use both testTag and content description?

Kazoroo replied:
Okay, I will change it.

sidcgithub commented on app/src/main/java/com/triviagenai/triviagen/trivia/presentation/answers/components/TriviaAnswerCard.kt:
```diff
@@ -50,7 +50,7 @@ fun TriviaAnswerCard(triviaQuestion: TriviaQuestion) {
             val checkedOption = 1 //TODO replace with state with real user answer. This value is for testing only.
 
-            for (i in 0..3) {
+            for (i in 0..triviaQuestion.options.lastIndex) {
```
Maybe using `forEachIndexed` would make this more readable

sidcgithub commented on app/src/main/java/com/triviagenai/triviagen/ui/theme/Type.kt:
```diff
@@ -35,6 +35,14 @@ val Typography = Typography(
+    titleMedium = TextStyle(
+        fontFamily = karmaFontFamily,
+        fontWeight = FontWeight.W600,
+        fontSize = 24.sp,
```
Would it be better to have `sp` in the `dimens` resource file?

Kazoroo replied:
No, this is the file containing text styles, and it is the only place in whole code where sp is used, so I think it is unnecesary.

sidcgithub reviewed (commented):
Nice! Left a few comments.

Kazoroo committed:
Resolve conflicts with merging viewmodels to navigation

Also add highlight of correct answer even if user answer wrong. Make option borders thicker when the option is wrong or correct.

Kazoroo committed:
Disable buttons in trivia while waiting for next question

Kazoroo committed:
Make for loop for TextButtons in MainMenu, move ButtonsData.kt to separate file, add quit app functionality

Kazoroo committed:
Change size for xlarge element. Add max points in TriviaScoreIndicator. Add padding to TextButton in MainMenuScreen.

sidcgithub commented on app/src/main/java/com/triviagenai/triviagen/trivia/presentation/TriviaQuestionViewModel.kt:
```diff
@@ -70,22 +79,27 @@ class TriviaQuestionViewModel @Inject constructor(
     fun processIntent(intent: TriviaIntent) {
         when (intent) {
-            is TriviaIntent.SubmitAnswer -> viewModelScope.launch { submitAnswer(intent.selectedOptionIndex) }
+            is TriviaIntent.SubmitAnswer -> viewModelScope.launch { submitAnswer(intent.selectedOptionIndex, intent.navController) }
```
I believe navigation should be tied to the UI and should not be executed from the ViewModel.

Could we use something like:
```    
LaunchedEffect(triviaRound.questions.size, triviaRound.currentQuestionIndex, questionBlock.selectedAnswer) {
        if (triviaRound.questions.isNotEmpty() &&
            triviaRound.questions.size - 1 == triviaRound.currentQuestionIndex &&
            questionBlock.selectedAnswer != SelectedAnswerState.Unanswered) {
            navController.navigate(Route.ResultsRoute) {
                popUpTo(Route.MainMenuRoute) {
                    inclusive = true
                }
            }
        }
    }
```
    
    This could be within the `DisplayGameContent` composable.

sidcgithub commented on app/src/main/java/com/triviagenai/triviagen/trivia/presentation/TriviaQuestionViewModel.kt:
```diff
@@ -70,22 +79,27 @@ class TriviaQuestionViewModel @Inject constructor(
     }
 
-    private fun nextQuestion() {
+    private fun nextQuestion(navController: NavHostController) {
```
`navController` can be removed.

sidcgithub commented on app/src/main/java/com/triviagenai/triviagen/trivia/presentation/TriviaQuestionViewModel.kt:
```diff
@@ -100,12 +114,18 @@ class TriviaQuestionViewModel @Inject constructor(
+
+
+    fun exitApp(activity: Activity) {
+        finishAffinity(activity)
```
I believe that the Activity should not be passed to the ViewModel. Maybe you can manage the state from the ViewModel and execute this in a Composable?

sidcgithub commented on app/src/main/java/com/triviagenai/triviagen/trivia/presentation/results/components/TriviaScoreIndicator.kt:
```diff
@@ -0,0 +1,40 @@
+    ) {
+        CircularProgressIndicator(
+            progress = score / (TriviaQuestionViewModel.POINTS * questionsSize),
+            color = if (score > 0.5f) TriviaGreen else TriviaRed,
```
I'm not sure I saw the colour as red when the score was below 0.5f
Score may need to be a remembered value. You could try something like:
`val score: Float by remember { mutableIntStateOf(0f) }`

sidcgithub commented on app/src/main/java/com/triviagenai/triviagen/trivia/presentation/roundsetup/RoundSetupScreen.kt:
```diff
@@ -50,21 +58,23 @@ fun RoundSetupScreen(triviaQuestionViewModel: TriviaQuestionViewModel) {
                 value = topicValue,
                 onValueChange = { topicValue = it },
-                label = { Text("Label") },
+                label = { Text("Trivia Topic") },
```
This can be moved to a string resource.

Kazoroo replied:
Done.

sidcgithub commented:
Hey I noticed that the screen recording was outdated. Could you please update it to the current functioning?

Kazoroo committed:
Change semantics modifier to testTag in composables

Kazoroo committed:
Use sealed class for enabling back navigation arrow

Kazoroo committed:
Move navigation after finished trivia to LaunchedEffect in DisplayGameContent from TriviaQuestionViewModel

Kazoroo committed:
Extract string resource, optimize imports

Kazoroo committed:
Extract string resource, optimize imports

Kazoroo committed:
Add placeholder for IP address

sidcgithub reviewed (approved):
Please move`fun exitApp(activity: Activity)` outside the ViewModel before squash merge.

Kazoroo committed:
Move app exit code to the MainMenuScreen

Also optimize imports and remove unused parameters
